ui/src/ducks/app/solutions.js

```diff
@@ -204,16 +204,41 @@ export function* updateEnvironments(environments) {
           env.name,
         );
         if (!solutionOperatorDeployment.error && !solutionUIDeployment.error) {
+          const currentSolutionVersion = envConfig?.[solution];
+          // when update the environment, we add `availableUpgradeVersion` and `availableDowngradeVersion` field
```

Comment:
We shouldn't do this kind of thing, it's duplicating existing data we have in the state. We should just make sure whenever it's consumed that we know how to compute these lists from a selected version and all the available ones (that we have stored in the state).

So, the code you wrote here should either go directly in the presentation component, or as a utility method. But the computation isn't expensive, so there's no need to cache the result in Redux.

Reply:
Hello @gdemonet,

So what you mean is to store `availableUpgradeVersion` and `availableDowngradeVersion` data in the local component state instead of caching it in redux store?

Reply:
Yeah, that's the idea. A single row can compute whatever it needs from its props and the list of available solutions in the global state :)